### PR TITLE
[SYCL] Fix the sub group size of Intel

### PR DESCRIPTION
**Pull request by luoyu-intel**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

## Changes:
1. sync the WARP_SIZE macro, and replace hard-coding 32 numbers.
2. add a macro condition to change WARP_SIZE to 16 for Intel's GPUs
3. remove hard-coding work_group_size=1024 constraint of *_norm_f32
4. remove unused shared local memory 
5. output the correct tokens for Debug build, before the logits are all `-nan`. It's an issue from dpcpp: https://github.com/intel/llvm/issues/14274
6. move *_norm_f32 to norm.cpp file
7. fix the bug of ignoring src1_ncols  of mmvq

Debug can output the same tokens as release in this PR(master runs into exceptions):
```
 What's your name?ggml_gallocr_needs_realloc: node inp_embd is not valid
ggml_gallocr_alloc_graph: cannot reallocate multi buffer graph automatically, call reserve
ggml_backend_sched_alloc_splits: failed to allocate graph, reserving

 Unterscheidung between the two words "name" and "namee"
```
Release output:
```
What's your name?
 Unterscheidung between the two words "name" and "namee"
```

## Performance benefit
Intel Arc A770
37 tokens/s to 39 tokens/s (Windows + 9600K):
```
 Building a website can be done in 10 simple steps: Choosing a hosting platform before you begin anything, though, should come first.
llama_print_timings:        load time =    5109.68 ms
llama_print_timings:      sample time =       0.80 ms /    16 runs   (    0.05 ms per token, 20025.03 tokens per second)
llama_print_timings: prompt eval time =     271.43 ms /    14 tokens (   19.39 ms per token,    51.58 tokens per second)
llama_print_timings:        eval time =     383.75 ms /    15 runs   (   25.58 ms per token,    39.09 tokens per second)
llama_print_timings:       total time =     666.89 ms /    29 tokens
Log end
```
38.9 tokens/s to 41.8 tokens/s (Linux + Xeon 4th):
```
 Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun in the way that little girls like you usually have. So she played outside with
llama_print_timings:        load time =   14153.51 ms
llama_print_timings:      sample time =       0.72 ms /    16 runs   (    0.04 ms per token, 22284.12 tokens per second)
llama_print_timings: prompt eval time =     391.66 ms /    33 tokens (   11.87 ms per token,    84.26 tokens per second)
llama_print_timings:        eval time =     359.29 ms /    15 runs   (   23.95 ms per token,    41.75 tokens per second)
llama_print_timings:       total time =     757.59 ms /    48 tokens
```

### CODE REVIEW

File: ggml-sycl/presets.hpp [hunk 1]
```diff
@@ -18,7 +18,11 @@
 #define GGML_SYCL_MAX_DEVICES       48
 #define GGML_SYCL_NAME "SYCL"
 
+#ifdef GGML_SYCL_NV_ARCH
 #define WARP_SIZE 32
+#else
+#define WARP_SIZE 16
+#endif
```
Comment (airMeng) [inline ggml-sycl/presets.hpp:25]: @AidanBeltonS here is for the issue of dpcpp mentioned in PR description

File: ggml-sycl/common.hpp [hunk 2]
```diff
@@ -294,5 +294,60 @@ struct ggml_backend_sycl_context {
...
+static inline int get_work_group_size(const sycl::device& device) {
+    dpct::device_info prop;
+    dpct::get_device_info(prop, device);
+    return prop.get_max_work_group_size();
+}
+
+
+// common device functions
+
+static __dpct_inline__ float warp_reduce_sum(float x,
```
Comment (airMeng) [inline ggml-sycl/common.hpp:308]: can we separate ```warp_reduce_xxx``` to independent files?

**Review by airMeng — APPROVED** [review]
need confirmation from our codeplay mates to verify on other HW

File: ggml-sycl/common.hpp (hunk 2 above)
Comment (luoyu-intel) [reply ggml-sycl/common.hpp:308]: no, there are inline functions. there are only little codes in common.hpp.

**Comment by airMeng** [comment]
@AidanBeltonS  @OuadiElfarouki @joeatodd if no regression on your side, I will merge it soon

**Comment by luoyu-intel** [comment]
The quantize functions limit the WARP_SIZE equals block size=32, there is remaining work for this.

File: ggml/src/ggml-sycl/presets.hpp [hunk 3]
```diff
@@ -16,7 +16,11 @@
 #define GGML_SYCL_MAX_STREAMS       8
 #define GGML_SYCL_MAX_BUFFERS       256
 
+#ifdef GGML_SYCL_NV_ARCH
 #define WARP_SIZE 32
+#else
+#define WARP_SIZE 16
+#endif
```
Comment (joeatodd) [inline ggml/src/ggml-sycl/presets.hpp:23]: At present GGML_SYCL_NV_ARCH is never defined so I had to patch this to be able to build. I think it would be useful to hook this up to `ggml/src/CMakeLists.txt`, where we have e.g.:

```CMake
        if (GGML_SYCL_TARGET STREQUAL "INTEL")
           #...
        elseif (GGML_SYCL_TARGET STREQUAL "NVIDIA")
           #...
        endif()
```

Secondly, I think many (most?) Intel platforms support both 32 and 16 for sg_size, so maybe you want to make it a CMake switch? For Codeplay's perspective, that is less important since we'll always use 32. Just a suggestion

**Review by joeatodd — CHANGES_REQUESTED** [review]
Hey, changes look good, nice to see some more work to split `ggml-sycl.cpp` into more small files! :+1:

Tested on our side, all fine except for the 2 small things I've raised.

File: ggml/src/ggml-sycl/presets.hpp (hunk 3 above)
Comment (luoyu-intel) [reply ggml/src/ggml-sycl/presets.hpp:23]: It seems that GGML_SYCL_NV_ARCH was removed by my previous rebase. I will add it back later.
I think 32 is just for compatibility. For the current Intel archs, 32 has no performance advantages but bugs. Using 16 only is okay for now.

File: ggml/src/ggml-sycl/mmvq.cpp [hunk 4]
```diff
@@ -948,77 +948,82 @@ void ggml_sycl_op_mul_mat_vec_q(
...
             break;
-    }
+        }
 
+    }
     (void) src1;
     (void) dst;
     (void) src1_ddf_i;
-    (void) src1_ncols;
     (void) src1_padded_row_size;
```
[the remark this answers is not in the document]
Comment (luoyu-intel) [reply ggml/src/ggml-sycl/mmvq.cpp:1028]: It's the wrong code from the quantize-correction commit. I'm still working on it as I found SYCL results are very different from CPU results.

[the remark this answers is not in the document]
Comment (luoyu-intel) [reply ggml/src/ggml-sycl/mmvq.cpp]: already fixed

File: ggml/src/ggml-sycl/presets.hpp (hunk 3 above)
Comment (luoyu-intel) [reply ggml/src/ggml-sycl/presets.hpp]: add back the macro definition.

**Comment by luoyu-intel** [comment]
Hi @joeatodd  please review the change.

**Comment by luoyu-intel** [comment]
I've fixed the src1_ncols bug of mmvq. But there is a remaining accuracy bug when the prompt length is less than 9.

This PR branch with prompt_length=9:
 **Once upon a time, there is a** small, remote village nestled among the rolling hills of the countryside
prompt_length=8:
**Once upon a time, there is** gepubliceerd a new version of the game Civilization IV called Civilization IV: Col

The master branch with prompt_length=9:
**Once upon a time, there is a** small village nestled in the mountains. Unterscheidung between the two is not always clear
prompt_length=8:
**Once upon a time, there is**iech shut shut shut shut shutoni oni  shut shut shutoni oni

**Review by joeatodd — APPROVED** [review]
Looks good, thanks for the changes :smiley:

File: ggml/src/ggml-sycl.cpp [hunk 5]
```diff
@@ -444,126 +358,11 @@ static void pad_f32(const float  *x, float *dst, const int ne0, const int ne00,
...
-
-    const float mean = tmp / ncols;
-    const float scale = sycl::rsqrt(mean + eps);
-
-    for (int col = tid; col < ncols; col += block_size) {
-        dst[row*ncols + col] = scale * x[row*ncols + col];
-    }
-}
-
+template<int QUANT_BLOCK_TILE>
```
[the remark this answers is not in the document]
Comment (luoyu-intel) [reply ggml/src/ggml-sycl.cpp:361]: This function is for Q8_1, so therer is only one fixed block size.

[the remark this answers is not in the document]
Comment (airMeng) [reply ggml/src/ggml-sycl.cpp:361]: yes, shall we add some check like
```
static_assert(QUANT_BLOCK_TILE ==32)
```

[the remark this answers is not in the document]
Comment (luoyu-intel) [reply ggml/src/ggml-sycl.cpp:361]: `QUANT_BLOCK_TILE = QK8_1/WARP_SIZE. ` It's not necessary to add this check, the kernel works for all `BlockSize%WARP_SIZE==0`.

**Comment by qnixsynapse** [comment]
I somehow missed this. Using this patch, the gemma model is broken, atleast on Q4_K_S

image

**Comment by luoyu-intel** [comment]
@qnixsynapse Is your prompt "Hi"? The SYCL backend had this repeat issue a long time ago.

**Comment by qnixsynapse** [comment]
This is llama -3 8B. Not sure what went wrong but speed has been increased.
image

**Comment by luoyu-intel** [comment]
> I've fixed the src1_ncols bug of mmvq. But there is a remaining accuracy bug when the prompt length is less than 9.
> 
> This PR branch with prompt_length=9: **Once upon a time, there is a** small, remote village nestled among the rolling hills of the countryside prompt_length=8: **Once upon a time, there is** gepubliceerd a new version of the game Civilization IV called Civilization IV: Col
> 
> The master branch with prompt_length=9: **Once upon a time, there is a** small village nestled in the mountains. Unterscheidung between the two is not always clear prompt_length=8: **Once upon a time, there is**iech shut shut shut shut shutoni oni shut shut shutoni oni
> 
> The main branch is worse than this PR. And I can tell that this is not related to the sub-group size. So I will fix it in the next PR instead of this one.

You can check this comment. How about a longer prompt?

**Comment by qnixsynapse** [comment]
llama-8B on a longer prompt. I have Arc A750 GPU if that matters.

image

**Comment by luoyu-intel** [comment]
I think there are two issues: a). short prompt produces the repeating tokens. b). garbage tokens when the context length is larger than some values.

**Comment by luoyu-intel** [comment]
@qnixsynapse The first one is confirmed as an existing issue of the master branch. I will look into the second one to see whether it is introduced by this PR.

**Comment by qnixsynapse** [comment]
It's a regression since before this patch it used to work well(although a bit slower). I am still trying to debug. Sorry that I couldn't test it before because I was hooked in testing Gemma models.

**Comment by luoyu-intel** [comment]
I didn't test Q4_K_S models. I will test it on A770.

**Comment by qnixsynapse** [comment]
Yup confirmed. Works great on CPU. 
Tested iQ4_XS and Q4_K_S models.
image

File: ggml/src/CMakeLists.txt [hunk 6]
```diff
@@ -486,9 +486,11 @@ if (GGML_SYCL)
     add_compile_options(-I./) #include DPCT
 
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-narrowing")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
     if (GGML_SYCL_TARGET STREQUAL "NVIDIA")
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl-targets=nvptx64-nvidia-cuda")
+        add_compile_definitions(GGML_SYCL_WARP_SIZE=32)
+    else()
+        add_compile_definitions(GGML_SYCL_WARP_SIZE=16)
```
Comment (luoyu-intel) [inline ggml/src/CMakeLists.txt:493]: @qnixsynapse Could you change this size to 32 and test your model?

**Comment by luoyu-intel** [comment]
```
PR Q4_0, warp_size=32
 Once upon a time, there is a small village nestled in the rolling hills of the countryside. Unterscheidung between the two is not always clear-cut, and both terms are often used interchangeably. The village is home to a small population of people who live and work together in a close-knit community.

PR Q4_0, warp_size=16
 Once upon a time, there is a small, remote village nestled among the rolling hills of the countryside.rezzo The villagers of the village were known for their exceptional craftsmanship and artistic abilities. They were skilled in the art of woodworking, weaving, and pottery. The villagers were also

PR Q4K_S, warp_size=32
 Once upon a time, there is a small village nestled in the mountains. The villagers lived simple lives, farming the land and raising their families. But one day, a great evil descended upon the village, in the form of a powerful sorcerer.
The sorcerer was angry and resentful towards the villagers, and

PR Q4K_S, warp_size=16
 Once upon a time, there is a smalloshtztztzrtrtrtrtttt tt tt tuleuleuleuleuleuleuleuleuleuleuleule Roman Roman Roman Roman Roman Romananeaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneaneane@{ane
```

I've confirmed this bug on Q4K_S.

File: ggml/src/CMakeLists.txt (hunk 6 above)
Comment (qnixsynapse) [reply ggml/src/CMakeLists.txt:493]: Works perfectly on warp size of 32!!!! 
image

**Comment by qnixsynapse** [comment]
> PR Q4_0, warp_size=16
 Once upon a time, there is a small, remote village nestled among the rolling hills of the countryside.rezzo The villagers of the village were known for their exceptional craftsmanship and artistic abilities. They were skilled in the art of woodworking, weaving, and pottery. The villagers were also

> hills of the countryside.rezzo

It's also broken on Q4_0

**Comment by airMeng** [comment]
@qnixsynapse BTW which UI are you using, looking quite cool

**Comment by qnixsynapse** [comment]
@airMeng It's chainlit.

**Comment by luoyu-intel** [comment]
@qnixsynapse WARP_SIZE=32 works fine for me. I can change WARP_SIZE to 32 for Intel GPUs in the new PR to revert this regression. Do you agree with this?

**Comment by qnixsynapse** [comment]
@luoyu-intel Sure. :)
